src/Microsoft.VisualStudio.Composition/Configuration/SerializationContextBase.cs

```diff
@@ -677,16 +672,48 @@ protected ImmutableArray<T> ReadImmutableArray<T>(BinaryReader reader, Func<T> i
 
                 if (count == 0)
                 {
-                    return ImmutableArray<T>.Empty;
+                    return Array.Empty<T>();
                 }
 
-                ImmutableArray<T>.Builder list = ImmutableArray.CreateBuilder<T>((int)count);
-                for (int i = 0; i < count; i++)
+                var list = new T[count];
+                for (int i = 0; i < list.Length; i++)
                 {
-                    list.Add(itemReader());
+                    list[i] = itemReader();
                 }
 
-                return list.MoveToImmutable();
+                return list;
+            }
+        }
+
+        protected ImmutableArray<T> ReadImmutableArray<T>(BinaryReader reader, Func<T> itemReader)
+        {
+            using (this.Trace(typeof(T).Name, isArray: true))
+            {
+                uint count = this.ReadCompressedUInt();
+
+                switch (count)
+                {
+                    case 0:
+                        return ImmutableArray<T>.Empty;
+                    case 1:
+                        return ImmutableArray.Create(itemReader());
+                    case 2:
+                        return ImmutableArray.Create(itemReader(), itemReader());
+                    case 3:
+                        return ImmutableArray.Create(itemReader(), itemReader(), itemReader());
+                    case 4:
+                        return ImmutableArray.Create(itemReader(), itemReader(), itemReader(), itemReader());
+                    default:
+                        if (count > 0xffff)
+                        {
+                            // Probably either file corruption or a bug in serialization.
+                            // Let's not take untold amounts of memory by throwing out suspiciously large lengths.
+                            throw new NotSupportedException();
+                        }
+
+                        IEnumerable<T> items = Enumerable.Range(1, (int)count).Select(i => itemReader());
```

Comment:
I don't like this, let me try caching a list of builders instead.

Reply:
Cleaned it up to not use Enumerable.Range anymore (this had a couple issues outlined in the 2nd commit description).

Reply:
Can you update the PR description to match the final iteration?